guppylang-internals/src/guppylang_internals/tys/parsing.py

```diff
@@ -330,13 +363,42 @@ def parse_parameter(
                 return TypeParam(
                     idx, node.name, must_be_copyable=True, must_be_droppable=True
                 )
-            # Otherwise, it must be a const parameter
+            # Otherwise, it must be either a protocol or a const parameter
             case bound:
+                from guppylang_internals.definition.protocol import ParsedProtocolDef
+
+                ctx = TypeParsingCtx(globals, param_var_mapping, allow_free_vars)
+                # First, try to see if this is a protocol bound by checking if can find
+                # a protocol definition with this name. In contrast to normal
+                # parameters, protocol parameters could be parametrised themselves.
+                proto_defn = None
+                proto_args = []
+                if isinstance(bound, ast.Subscript):
+                    proto_defn = _try_parse_defn(bound.value, ctx.globals)
+                    arg_nodes = (
+                        bound.slice.elts
+                        if isinstance(bound.slice, ast.Tuple)
+                        else [bound.slice]
+                    )
+                    proto_args = [arg_from_ast(arg_node, ctx) for arg_node in arg_nodes]
+                else:
+                    proto_defn = _try_parse_defn(bound, ctx.globals)
+                if isinstance(proto_defn, ParsedProtocolDef):
+                    if proto_args == []:
+                        proto_args = [param.to_bound() for param in proto_defn.params]
```

Comment:
Don't we need to add these new params to the `param_var_mapping`?

Also, we probably need to specify a manual de Bruijn index, i.e. `param.to_bound(idx=len(ctx.param_var_mapping))`?